### PR TITLE
CATROID-915 Separation of Visual Sensors into Face & Text Sensors

### PR DESCRIPTION
**Pull request by ISKalsi**
Ticket: https://jira.catrob.at/browse/CATROID-915

Before | After
-------| ----- 
 | 

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [x] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [x] Stick to the project’s gitflow workflow
- [x] Verify that your changes do not have any conflicts with the base branch
- [x] After the PR, verify that all CI checks have passed
- [ ] Post a message in the *catroid-stage* or *catroid-ide* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

**Comment by ISKalsi** [comment]
> Hello,
> thank you very much for this PR! Unfortunately, we missed your comment on that ticket, I apologize for the long waiting time for this review. However, we now blocked it for you on jira and I finally reviewed it and I have a view remarks:
> 
> 1. Please rename your branch to CATROID-915 to match our naming convention.
> 2. Please only change the 'catroid/src/main/res/values/strings.xml' because every other string.xml will be overwritten by our translations automatically.
> 3. Please use the strings that are specified by the [ticket](https://jira.catrob.at/browse/CATROID-915)
> 4. Resolve the [Warning](https://jenkins.catrob.at/job/Catroid/job/PR-3981/2/analysis/source.c0f74023-5f38-451b-8406-7893b70e2739/#641) on Jenkins
> 5. implement a reasonable test for your implemention
> 
> Please also **rebase** on the current upstream develop, because there are several merge conflicts! Also, keep in mind that there should be only one commit per PR, so please squash any additional commits together.
> Best Regards!

I rebased on the develop branch, but there are still merge conflicts that need to be resolved. I'm not sure what I did wrong. Here is what I did.

```
git checkout CATROID-915
git rebase -i origin/develop

# conflicts arrived, merged them using - 
git mergetool

git rebase --continue
git push origin CATROID-915 --force
```

It would be great if you could suggest me what went wrong
Thank you!

**Comment by Walcho1125** [comment]
Hello, you need to rebase onto the upstream/develop, your latest commit on the origin/develop and on this branch is apparently from 8.2.2021 and the conflict that occurs here is from a commit that was made 20.2.2021, 
That should resolve the problem
BR

**Comment by ISKalsi** [comment]
Hello, Jenkins is giving this warning in the strings.xml file. I think it may be a mistake as similar names exist which are not showing in the warning. Should I go ahead and remove the resource or just ignore it?
Thank you!

Screenshot 2021-03-26 at 1 36 44 AM

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/CategoryListFragment.java [hunk 1]
```diff
@@ -741,9 +726,12 @@ private void initializeAdapter() {
...
 
-	private List<CategoryListItem> getVisualSensorItems() {
-		return addHeader(toCategoryListItems(SENSORS_VISUAL_DETECTION, SENSORS_VISUAL_DETECTION_PARAMS),
-				getString(R.string.formula_editor_device_visual_detection));
+	private List<CategoryListItem> getFaceSensorItems() {
+		return addHeader(toCategoryListItems(SENSORS_FACE_DETECTION), getString(R.string.formula_editor_device_face_detection));
+	}
+
+	private List<CategoryListItem> getTextSensorItems() {
+		return addHeader(toCategoryListItems(SENSORS_TEXT_DETECTION), getString(R.string.formula_editor_device_text_detection));
```
Comment (Walcho1125) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/CategoryListFragment.java:734]: You should NOT just overwrite the changes with yours, they were added for a reason, please add the SENSORS_VISUAL_DETECTION_PARAMS also to these items accordingly(split them into 2 Lists).
This should also fix the Warning on Jenkins

**Comment by Walcho1125** [comment]
Please make sure that all tests in the catrobat testrunner run locally, as well as on Jenkins, 
Thanks and best regards!

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/CategoryListFragment.java (hunk 1 above)
Comment (ISKalsi) [reply catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/CategoryListFragment.java:734]: Ohh. Okay I'm sorry, will revert that.

**Comment by ISKalsi** [comment]
> Please make sure that all tests on the catrobat testrunner run locally, as well as on Jenkins, 
> 
> Thanks and best regards!

Regarding that, i had a doubt. I wrote 2 tests for my implementation but they both require camera. So i was thinking to create a brick in the "device" category that mocks a videostream. Or is there anything that already exists that can be done?

**Comment by ISKalsi** [comment]
Hello, All the tests that are failing on Jenkins, are passing locally. What should I do?
Sorry for the inconvenience caused by my questions. 😓

**Comment by Walcho1125** [comment]
> Hello, All the tests that are failing on Jenkins, are passing locally. What should I do?
> Sorry for the inconvenience caused by my questions. sweat

Hello, No Problem, sorry I had to take a closer look into that problem, it seems like one of the catrobat tests fails in a way, that makes the other tests also fail afterward. It seems that one of your tests causes that problem. Please try to fix your test, to find out which one causes the problem you could remove one after another and see what Jenkins does. 
BR

**Comment by ISKalsi** [comment]
You were right. Thank you! All tests passed but one, will try and fix it asap.

**Comment by ISKalsi** [comment]
Hello, tried fixing that one failing test, spent the whole night trying to fix many other that followed :'(
I think we can raise issues for them separately; it seems like a very tedious job to get them all right at the same time. 
Thanks for helping, please review the updated PR.

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/CategoryListFragment.java [hunk 2]
```diff
@@ -198,36 +198,37 @@
...
-			R.string.formula_editor_function_text_block_parameter,
-			R.string.formula_editor_function_text_block_parameter,
-			R.string.formula_editor_function_text_block_parameter,
-			R.string.formula_editor_function_text_block_parameter);
+					R.string.formula_editor_function_no_parameter,
+					R.string.formula_editor_function_text_block_parameter,
+					R.string.formula_editor_function_text_block_parameter,
+					R.string.formula_editor_function_text_block_parameter,
+					R.string.formula_editor_function_text_block_parameter,
+					R.string.formula_editor_function_text_block_parameter);
```
Comment (Walcho1125) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/CategoryListFragment.java:231]: please remove additional intend

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/CategoryListFragment.java [hunk 3]
```diff
@@ -198,36 +198,37 @@
...
+			R.string.formula_editor_sensor_second_face_y_position);
+	private static final List<Integer> SENSORS_FACE_DETECTION_PARAMS =
+			asList(R.string.formula_editor_function_no_parameter,
+					R.string.formula_editor_function_no_parameter,
+					R.string.formula_editor_function_no_parameter,
+					R.string.formula_editor_function_no_parameter,
+					R.string.formula_editor_function_no_parameter,
+					R.string.formula_editor_function_no_parameter,
+					R.string.formula_editor_function_no_parameter,
+					R.string.formula_editor_function_no_parameter);
```
Comment (Walcho1125) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/CategoryListFragment.java:216]: please remove additional intend

**Review by Walcho1125 — CHANGES_REQUESTED** [review]
Thanks for your effort, it seems to work now! 
However please please remove the additional intend as stated in the comments to improve readability

**Comment by ISKalsi** [comment]
Done. That inconsistent indent was a good find! :D

**Review by Walcho1125 — APPROVED** [review]
Well done, 
Thank you!

**Review by wslany — CHANGES_REQUESTED** [review]
Thanks a lot!  A few things: 

- [ ] Capitalization, see my change request in the code.
- [ ] Most asserts in the Catrobat language test are always true, e.g., the following simply because x = 100 + x - 100. They thus do not test the correct functionality of the sensors (and they also do not test the existence of the sensors in the list in the interface, e.g., extension sensors that are not shown unless activated in the settings but can be used in formulas).
image
or, in 2d script-view:
image
- [ ] The two Catrobat language tests seemingly were written with the intent to check the _functionality_ of the sensors, but the PR does not implement any code for their functionality, and as such already these tests would not belong to this ticket. The visual sensor _functionality_ **cannot** be tested in such a simple way. There _are_ ways to test them, but those tests would be much more complex, and you don't need to include any such tests anyway since it is not the topic of this ticket. Please remove both tests.
- [ ] However, no tests have been included in this pull request that check whether the sensors can now be _found_ in the menu below the new headers and in the right order, and that the old header has been removed. Please add such tests. 
- [ ] If I may ask you, the current order of all the sensor categories is not ideal, as the sensor categories are not ordered by their usefulness and most likely frequency the users may want to use the sensors in their projects. Note that the sensors for extensions such as Arduino will only be shown if the corresponding extension is activated in the settings on the main page of the app, and it therefore would be better to place these sensors at the top of the list, where they will be shown more prominently to users who consciously have activated such an extension (we do the same for brick categories of extensions already, e.g., when the embroidery extension is activated, which will be shown at the top of the brick categories when activated):  

1. NXT sensor sensors 
2. EV3 sensors 
3. Phiro sensors 
4. Arduino sensors 
5. Drone sensors 
6. Raspberry Pi sensors 
7. NFC sensors 
8. Cast gamepad sensors 
9. Device sensors 
10. Touch sensors 
11. Date and time sensors 
12. Speech recognition sensors 
13. Face sensors 
14. Text sensors

**Comment by ISKalsi** [comment]
>  The two Catrobat language tests seemingly were written with the intent to check the _functionality_ of the sensors

No actually, I wrote them with the intent to test their existence only, as you also said earlier, all of the assert bricks are actually always true. But yes I understand how _that_ also is not fulfilling the requirement. 
Will take care of that and all other changes asap.
Thanks!

**Comment by ISKalsi** [comment]
* [ ]  Changed capitalisation
* [ ]  Removed both the catrobat tests
* [ ]  Changed the order of sensors according to the your requirement
> 1. NXT sensor sensors
> 2. EV3 sensors
> 3. Phiro sensors
> 4. Arduino sensors
> 5. Drone sensors
> 6. Raspberry Pi sensors
> 7. NFC sensors
* [ ]  Implemented a reasonable test (hopefully) for all of the above. (Please pardon my mistake. This is my first ever test)

**Review by Walcho1125 — CHANGES_REQUESTED** [review]
Sorry for the late review, I missed your pushes.
It looks good, however, we follow strict clean code guidelines which require no comments in the code. Please remove the comments in FormulaEditorSensorListTest.kt as the names already give away their purpose.
Thanks!

**Comment by ISKalsi** [comment]
Okay. Also sir, I had a question, many of such instrumentation tests are marked with a category (additional to `Cat.AppUI`) which is `Level.Smoke`. I couldn't figure out which group of tests does it classify.

**Comment by Walcho1125** [comment]
> Also sir, I had a question, many of such instrumentation tests are marked with a category (additional to `Cat.AppUI`) which is `Level.Smoke`. I couldn't figure out which group of tests does it classify. Should I add it in this test too?
> Thanks!

As it is also in the other FormulaEditorTests please add it, thanks for the note

**Review by Walcho1125 — APPROVED** [review]
Well done, looks good to me!

**Review by wslany — APPROVED** [review]

**Comment by wslany** [comment]
Thanks a lot!

**Pull request merged**